services/QuillLMS/client/app/bundles/Comprehension/components/studentView/container.tsx

```diff
@@ -92,12 +101,62 @@ export class StudentViewContainer extends React.Component<StudentViewContainerPr
     }
 
     window.addEventListener('keydown', this.handleKeyDown)
+    window.addEventListener('mousemove', this.resetTimers)
+    window.addEventListener('mousedown', this.resetTimers)
+    window.addEventListener('click', this.resetTimers)
+    window.addEventListener('keypress', this.resetTimers)
+    window.addEventListener('visibilitychange', this.setIdle)
   }
 
   componentWillUnmount() {
     window.removeEventListener('keydown', this.handleKeyDown)
+    window.removeEventListener('mousemove', this.resetTimers)
+    window.removeEventListener('mousedown', this.resetTimers)
+    window.removeEventListener('click', this.resetTimers)
+    window.removeEventListener('keypress', this.resetTimers)
+    window.removeEventListener('visibilitychange', this.setIdle)
+  }
+
+  outOfAttemptsForActivePrompt = () => {
+    const { activeStep, } = this.state
+    const { session, } = this.props
+    const { submittedResponses, } = session
+    const activePrompt = this.orderedSteps()[activeStep - 2] // subtracting 2 because the READ_PASSAGE_STEP is 1, so the first item in the set of prompts will always be 2
+
+    if (!activePrompt) { return }
+
+    const responsesForPrompt = submittedResponses[activePrompt.id]
+
+    if (!responsesForPrompt) { return }
+
+    const lastAttempt = responsesForPrompt[responsesForPrompt.length - 1]
+
+    return (responsesForPrompt.length === activePrompt.max_attempts) || lastAttempt.optimal
   }
 
+  resetTimers = (e=null) => {
+    this.setState((prevState, props) => {
+      const { activeStep, startTime, timeTracking, isIdle, inactivityTimer, completedSteps, } = prevState
+      if (completedSteps.includes(activeStep)) { return } // don't want to add time if a user is revisiting a previously completed step
+      if (this.outOfAttemptsForActivePrompt()) { return }// or if they are finished submitting responses for the current active step
+
+      if (inactivityTimer) { clearTimeout(inactivityTimer) }
+
+      let elapsedTime = _.round((Date.now() - startTime) / 1000)
+      if (isIdle) {
+        elapsedTime = 0
+      }
+      const newTimeTracking = {...timeTracking, [activeStep]: timeTracking[activeStep] + elapsedTime}
+      const newInactivityTimer = setTimeout(this.setIdle, 30000);  // time is in milliseconds (1000 is 1 second)
+
+      return { timeTracking: newTimeTracking, isIdle: false, inactivityTimer: newInactivityTimer, startTime: new Date(), }
```

Comment:
This is minor enough to not actually matter for practical purposes, but I figured I'd call it out for completeness sake: we use `Date.now()` for the elapsedTime calculation, and then we use it again here as a new call for `startTime`.  Technically, some very small number of ms (1 or 2, maybe?) will often pass between these two calls, meaning that we "lose" the time between the two invocations.  1 or 2 ms, even if we were tracking across thousands of events, probably doesn't matter to us, but it might be worth refactoring to call `Date.now()` once at the top of this function and use the result obtained there for both the `elapsedTime` calculation and the new value of `startTime`.

If you feel like it.

Reply:
Good point. I'll update.